File: Server/client/ogn/Assets/script/shared/BinaryStream.cs
using System.Collections;
using System.IO;
using System;
using System.Text;

public class BinaryStream 
{
    public const int PACKET_MAX_LENGTH = 4096;
    public static bool HostEqualNet()
    {
        short v = 1;
        byte[] data = BitConverter.GetBytes(v);
        if (data[0] == 1) return false;
        return true;             
    }

    public static int ntohl(int value)
    {
        if (HostEqualNet())
            return value;

        byte[] data = NetToHost(BitConverter.GetBytes(value));

        return BitConverter.ToInt32(data, 0);
    }

    //将主机顺序转换为网络顺序;
    public static byte[] HostToNet(byte[] data)
    {
        if (HostEqualNet())
            return data;

        byte[] temp = new byte[data.Length];
        for (int i = 0; i < data.Length; ++i )
            temp[i] = data[data.Length - i - 1];

        return temp;
    }

    //将网络顺序转换为主机顺序;
    public static byte[] NetToHost(byte[] data)
    {
        byte[] temp = new byte[data.Length];
        for (int i = 0; i < data.Length; ++i)
            temp[i] = data[data.Length - i - 1];

        return temp;
    }

    public BinaryStream ()
    {
        mBuffer = new byte[Shared.MAX_PACKET_LENGTH];
        mWriteIndex = 0;
        mReadIndex = 0;
    }

    public BinaryStream(int count)
    {
        mBuffer = new byte[count];
        mWriteIndex = 0;
        mReadIndex = 0;
    }

    public BinaryStream(byte[] data)
    {
        mBuffer = data;
        mWriteIndex = 0;
        mReadIndex = 0;
    }
    public BinaryStream(byte[] data, int offset, int length)
    {
        mBuffer = new byte[length];
        mWriteIndex = 0;
        mReadIndex = 0;

        Write(data, offset, length);
    }

    public bool Write(float value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(double value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(string value)
    {
        ushort leng = 0;
        byte[] data = null;
        if (value.Length > 0)
        {
            data = Encoding.UTF8.GetBytes(value);
            leng = (ushort)data.Length;
        }

        byte[] leng_data = HostToNet(BitConverter.GetBytes(leng));
        if (!Write(leng_data))
            return false;

        if (leng > 0)
            return Write(data);
       
        return true;
    }

    public bool Write(bool value)
    {
        validBuffer(1);
        mBuffer[mWriteIndex] = value ? (byte)1 : (byte)0;
        mWriteIndex += sizeof(byte);
        return true;
    }

    public bool Write(byte value)
    {
        validBuffer(1);
        mBuffer[mWriteIndex] = value;
        mWriteIndex += sizeof(byte);
        return true;
    }

    public bool Write(sbyte value)
    {
        validBuffer(1);
        mBuffer[mWriteIndex] = (byte)value;
        mWriteIndex += sizeof(sbyte);
        return true;
    }

    public bool Write(char value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(short value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(ushort value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(int value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(uint value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(long value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }

    public bool Write(ulong value)
    {
        byte[] data = HostToNet(BitConverter.GetBytes(value));
        if (!Write(data))
            return false;
        return true;
    }
    public bool Write(BinaryStream value)
    {
        if (!this.Write(value.wpos))
            return false;
        
        if (value.wpos <= 0)
            return true;
         
        if (!this.Write(value.buffer, 0, value.wpos))
            return false;
        
        return true;
    }

    public bool Write(Variant value)
    {
        Write(value.mType);
        switch (value.mType)
        {
            case Variant.TypeBoolean:
                return Write(value.GetBool());
            case Variant.TypeInt8:
                return Write(value.GetInt8());
            case Variant.TypeUint8:
                return Write(value.GetUint8());
            case Variant.TypeInt16:
                return Write(value.GetInt16());
            case Variant.TypeUint16:
                return Write(value.GetUint16());
            case Variant.TypeInt32:
                return Write(value.GetInt32());
            case Variant.TypeUint32:
                return Write(value.GetUint32());
            case Variant.TypeInt64:
                return Write(value.GetInt64());
            case Variant.TypeUint64:
                return Write(value.GetUint64());
            case Variant.TypeFloat32:
                return Write(value.GetFloat32());
            case Variant.TypeFloat64:
                return Write(value.GetFloat64());
            case Variant.TypeString:
                return Write(value.GetString());
        }
        return true;
    }

    public bool Write(byte[] buffer, int index = 0, int count = -1)
    {
        count = count == -1 ? buffer.Length : count;
        if (!validBuffer(count))
            return false;
        Array.Copy(buffer, index, mBuffer, mWriteIndex, count);
        mWriteIndex += count;
        return true;
    }

    public bool validBuffer(int count)
    {
        if (mBuffer.Length - mWriteIndex < count)
        {
            byte[] datas = new byte[mBuffer.Length + count];
            Array.Copy(mBuffer, 0, datas, 0, mBuffer.Length);
            mBuffer = datas;
            //Console.WriteLine("空间不足!");
            return true;
        }
        return true;
    }

    public bool validData(int count)
    {
        if (mBuffer.Length - mReadIndex < count)
        {
            Console.WriteLine("空间不足!");
            return false;
        }
        return true;
    }
 

    public bool Write(Header value)
    {
        if (!value.serialize(this))
            return false;
        return true;
    }

    public bool Read(ref float value)
    {
        int length = sizeof(float);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;
        value = BitConverter.ToSingle(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref double value)
    {
        int length = sizeof(double);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;
        value = BitConverter.ToDouble(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref string value)
    {
        int length = sizeof(short);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        short str_len = BitConverter.ToInt16(NetToHost(data), 0);

        if (str_len > 0)
        {
            byte[] str_data = new byte[str_len];
            if (!ReadBytes(str_data))
                return false;
            value = Encoding.UTF8.GetString(str_data, 0, str_len);
        }
        return true;
    }

    public bool Read(ref bool value)
    {
        if (!validData(sizeof(sbyte)))
            return false;
        value = mBuffer[mReadIndex] == 1 ? true : false;
        mReadIndex += sizeof(byte);
        return true;
    }

    public bool Read(ref byte value)
    {
        if (!validData(sizeof(sbyte)))
            return false;
        value = mBuffer[mReadIndex];
        mReadIndex += sizeof(byte);
        return true;
    }

    public bool Read(ref sbyte value)
    {
        if (!validData(sizeof(sbyte)))
            return false;
        value = (sbyte)mBuffer[mReadIndex];
        mReadIndex += sizeof(sbyte);
        return true;
    }

    public bool Read(ref char value)
    {
        int length = sizeof(char);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToChar(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref short value)
    {
        int length = sizeof(short);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToInt16(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref ushort value)
    {
        int length = sizeof(ushort);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToUInt16(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref int value)
    {
        int length = sizeof(int);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToInt32(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref uint value)
    {
        int length = sizeof(uint);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToUInt32(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref long value)
    {
        int length = sizeof(long);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToInt64(NetToHost(data), 0);
        return true;
    }

    public bool Read(ref ulong value)
    {
        int length = sizeof(ulong);

        byte[] data = new byte[length];
        if (!ReadBytes(data))
            return false;

        value = BitConverter.ToUInt64(NetToHost(data), 0);
        return true;
    }

    public bool ReadBytes(byte[] buffer, int index = 0, int length = -1)
    {
        length = length == -1 ? buffer.Length : length;
        if (!validData(length))
            return false;
        Array.Copy(mBuffer, mReadIndex, buffer, index, length);
        mReadIndex += length;

        return true;
    }


    public bool Read(Header value)
    {
        return value.deSerialize(this);
    }

    public bool Read(Variant value)
    {
        Read(ref value.mType);
        switch (value.mType)
        {
            case Variant.TypeBoolean:
                return Read(ref value.mValue.value_bool);
            case Variant.TypeInt8:
                return Read(ref value.mValue.value_int8);
            case Variant.TypeUint8:
                return Read(ref value.mValue.value_uint8);
            case Variant.TypeInt16:
                return Read(ref value.mValue.value_int16);
            case Variant.TypeUint16:
                return Read(ref value.mValue.value_uint16);
            case Variant.TypeInt32:
                return Read(ref value.mValue.value_int32);
            case Variant.TypeUint32:
                return Read(ref value.mValue.value_uint32);
            case Variant.TypeInt64:
                return Read(ref value.mValue.value_int64);
            case Variant.TypeUint64:
                return Read(ref value.mValue.value_uint64);
            case Variant.TypeFloat32:
                return Read(ref value.mValue.value_float32);
            case Variant.TypeFloat64:
                return Read(ref value.mValue.value_float64);
            case Variant.TypeString:
                return Read(ref value.mValue.value_string);
        }
        return true;
    }

    public bool Read(BinaryStream value)
    {
        int len = 0;
        this.Read(ref len);
        if (len >  0)
        {
            byte[] bytes = new byte[len];
            if (!ReadBytes(bytes))
                return false;
            value.Write(bytes, 0, len);       
        }
        return true;
    }

    public void Clear()
    {
        mWriteIndex = 0;
        mReadIndex = 0;
    }


    public byte[] buffer
    {
        get { return mBuffer; }
    }

    public int wpos
    {
        get { return mWriteIndex;  }
        set { mWriteIndex = value; }
    }

    public int rpos
    {
        get { return mReadIndex; }
        set { mReadIndex = value; }
    }

    byte [] mBuffer;
    int mWriteIndex;
    int mReadIndex;
}

File: Server/src/Shared/core/core/BinaryStream.cpp
#include "Shared.hpp"


BinaryStream::BinaryStream(int32 count /* = BINARY_LENGTH */):
mCount(count),
mBytes(NULL),
mWritePosition(0),
mReadPosition(0),
mIsDel(false)
{
	mIsDel = true;
	if (mCount <= 0)
		mBytes = NULL;
	else {
		mBytes = new int8[mCount];
		memset(mBytes, 0, mCount);
	}
}

BinaryStream::BinaryStream(void* ptr, int32 count, bool isDel):
mCount(count),
mBytes((int8*)ptr),
mWritePosition(0),
mReadPosition(0),
mIsDel(isDel)
{

}

BinaryStream::BinaryStream( const BinaryStream& other ):
mCount(other.mCount),
mWritePosition(other.mWritePosition),
mReadPosition(other.mReadPosition)
{
	mIsDel = true;
	mBytes = new int8[mCount];
	memcpy(mBytes, other.mBytes, mCount);
}

BinaryStream::~BinaryStream(void)
{
	release();
}

BinaryStream& BinaryStream::operator=( const BinaryStream& other )
{
	release();
	mCount = other.mCount;
	mWritePosition = other.mWritePosition;
	mReadPosition = other.mReadPosition;

	mIsDel = true;
	mBytes = new int8[mCount];
	memcpy(mBytes, other.mBytes, mCount);
	return *this;
}

void BinaryStream::clear()
{
	mWritePosition = 0;
	mReadPosition = 0;
}

std::string BinaryStream::toString()
{
	return Shared::ConvertHex((char*)datas(), wpos());
}

bool BinaryStream::write(const void* data, int32 count)
{
	if (count <= 0)
		return false;
	if (mWritePosition + count > mCount) {
		resize(mWritePosition + count);
	}

	memcpy(mBytes + mWritePosition, data, count);
	mWritePosition += count;

	return true;
}

bool BinaryStream::push(int32 pos, const void* data, int32 count)
{
	if (count <= 0)
		return false;

	if (pos + count > mCount)
		return false;
	memcpy(mBytes + pos, data, count);
	return true;		
}

bool BinaryStream::read(void* data, int32 count)
{
	if (count <= 0)
		return false;

	if (mReadPosition + count > mCount)
		return false;

	memcpy(data, mBytes + mReadPosition, count);
	mReadPosition += count;
	return true;
}

bool BinaryStream::pop(int32 pos, void* data, int32 count)
{
	if (count <= 0)
		return false;

	if (pos + count > mCount)
		return false;
	memcpy(data, mBytes + pos, count);
	return true;
}

void BinaryStream::resize(int32 size)
{
	if (size <= 0)
		return;
	mCount = size;
	int8* data = new int8[mCount];
	memcpy(data, mBytes, wpos());
	release();
	mIsDel = true;
	mBytes = data;
}

void BinaryStream::release()
{
	if (mIsDel)
		SAFE_DELETE_ARRAY(mBytes);
	mBytes = NULL;
}

bool BinaryStream::operator>>( Variant& value )
{
	int8 type_ = 0;
	CHECK(*this >> type_);
	switch (type_)
	{
	case Variant::TypeBoolean:
		{
			int8 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeInt8:
		{
			int8 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeUint8:
		{
			uint8 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeInt16:
		{
			int16 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeUint16:
		{
			uint16 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeInt32:
		{
			int32 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeUint32:
		{
			uint32 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeInt64:
		{
			int64 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeUint64:
		{
			uint64 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeFloat32:
		{
			float32 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeFloat64:
		{
			float64 v = 0;
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeString:
		{
			std::string v = "";
			CHECK(*this >> v);
			value = v;
			return true;
		}
	case Variant::TypeMemory:
		{
			BinaryStream& bytes = (BinaryStream&)value.valueBytes();
			CHECK(*this >> bytes);
			return true;
		}
	}
	return false;
}

bool BinaryStream::operator<<(const float32& value)
{
	float sv = Shared::htonf(value);

	return write(&sv, sizeof(value));
}

bool BinaryStream::operator<<( const float64& value )
{
	float64 sv = Shared::htond(value);

	return write(&sv, sizeof(value));
}

bool BinaryStream::operator<<(const std::string& value)
{
	int16 size = (int16)value.length();
	*this << size;
	if (size > 0)
		return write(value.c_str(), size);

	return true;
}

bool BinaryStream::operator<<(const char* value)
{
	int16 size = (int16)strlen(value);
	*this << size;
	if (size > 0)
		return write(value, size);

	return true;
}

bool BinaryStream::operator <<(const int8 value)
{
	return write(&value, sizeof(value));
}

bool BinaryStream::operator<<( const Variant& value )
{
	CHECK(*this << value.type());

	switch (value.type())
	{
	case Variant::TypeBoolean:
		CHECK(*this << value.valueBool());
		break;
	case Variant::TypeInt8:
		CHECK(*this << value.valueInt8());
		break;
	case Variant::TypeUint8:
		CHECK(*this << value.valueUint8());
		break;
	case Variant::TypeInt16:
		CHECK(*this << value.valueInt16());
		break;
	case Variant::TypeUint16:
		CHECK(*this << value.valueUint16());
		break;
	case Variant::TypeInt32:
		CHECK(*this << value.valueInt32());
		break;
	case Variant::TypeUint32:
		CHECK(*this << value.valueUint32());
		break;
	case Variant::TypeInt64:
		CHECK(*this << value.valueInt64());
		break;
	case Variant::TypeUint64:
		CHECK(*this << value.valueUint64());
		break;
	case Variant::TypeFloat32:
		CHECK(*this << value.valueFloat32());
		break;
	case Variant::TypeFloat64:
		CHECK(*this << value.valueFloat64());
		break;
	case Variant::TypeString:
		CHECK(*this << std::string(value.valueString()));
		break;
	case Variant::TypeMemory:
		BinaryStream& bytes = (BinaryStream&)((Variant&)(value)).valueBytes();
		CHECK(*this << bytes);
		break;
	}
	return true;
}

bool BinaryStream::operator<<(const uint8 value)
{
	return write(&value, sizeof(value));
}

bool BinaryStream::operator<<(const int16 value)
{
	int16 v = Shared::htons(value);
	
	return write(&v, sizeof(value));
}

bool BinaryStream::operator<<(const uint16 value)
{
	uint16 v = Shared::htons(value);
	
	return write(&v, sizeof(value));
}

bool BinaryStream::operator<<(const int32 value)
{
	int32 v = Shared::htonl(value);
	
	return write(&v, sizeof(value));
}

bool BinaryStream::operator<<(const uint32 value)
{
	uint32 v = Shared::htonl(value);
	
	return write(&v, sizeof(value));
}

bool BinaryStream::operator<<(const int64 value)
{
	int64 v = Shared::htonll(value);

	return write(&v, sizeof(value));
}

bool BinaryStream::operator<<(const uint64 value)
{
	uint64 v = Shared::htonll(value);
	
	return write(&v, sizeof(value));
}

bool BinaryStream::operator<<(const Object& value)
{
	Object& obj = (Object&)value;
	return obj >> (*this);
}

bool BinaryStream::operator<<(BinaryStream& value)
{
	int32 count = value.wpos();
	if (!(*this << count))
		return false;

	if (count <= 0)
		return true;

	return write(value.datas(), value.wpos());
}

bool BinaryStream::operator>>( float32& value )
{
	if (!read(&value, sizeof(float32)))
		return false;
	
	float32* v = (float32*)&value;
	*v = Shared::ntohf(*v);
	return true;
}

bool BinaryStream::operator>>(float64& value )
{
	if (!read(&value, sizeof(float64)))
		return false;

	float64* v = (float64*)&value;
	*v = Shared::ntohd(*v);

	return true;
}

bool BinaryStream::operator>>(std::string& value)
{
	int16 size = 0;
	if (!(*this >> size))
		return false;

	if (size <= 0)
		return true;

	char* str = new char[size + 1];
	if (!read(str, size))
	{
		delete[] str;
		return false;
	}

	str[size] = 0;
	value = str;
	delete[] str;
	return true;
}

bool BinaryStream::operator>>(int8& value )
{
	return read(&value, sizeof(int8));
}

bool BinaryStream::operator>>(uint8& value )
{
	return read(&value, sizeof(uint8));
}

bool BinaryStream::operator>>(int16& value )
{
	if (!read(&value, sizeof(int16)))
		return false;
	
	int16* v = (int16*)&value;
	*v = Shared::ntohs(*v);
	return true;
}

bool BinaryStream::operator>>(uint16& value )
{
	if (!read(&value, sizeof(uint16)))
		return false;

	uint16* v = (uint16*)&value;
	*v = Shared::ntohs(*v);

	return true;
}

bool BinaryStream::operator>>( int32& value )
{
	if (!read(&value, sizeof(int32)))
		return false;

	int32* v = (int32*)&value;
	*v = Shared::ntohl(*v);

	return true;
}

bool BinaryStream::operator>>(uint32& value )
{
	if (!read(&value, sizeof(uint32)))
		return false;

	uint32* v = (uint32*)&value;
	*v = Shared::ntohl(*v);

	return true;
}

bool BinaryStream::operator>>(int64& value )
{
	if (!read(&value, sizeof(int64)))
		return false;

	int64* v = (int64*)&value;
	*v = Shared::ntohll(*v);

	return true;
}

bool BinaryStream::operator>>( uint64& value )
{
	if (!read(&value, sizeof(uint64)))
		return false;

	uint64* v = (uint64*)&value;
	*v = Shared::ntohll(*v);

	return true;
}

bool BinaryStream::operator>>(Object& value)
{
	return value << (*this);
}

bool BinaryStream::operator>>(BinaryStream& value)
{
	int bit_size = 0;
	if (!(*this >> bit_size))
		return false;

	if (bit_size <= 0)
		return true;
	
	int8* bit_data = new int8[bit_size];
	if (!read(bit_data, bit_size))
	{
		delete [] bit_data;
		return false;
	}

	if (!value.write(bit_data, bit_size))
	{
		delete [] bit_data;
		return false;
	}

	delete [] bit_data;
	return true;	 
}
